### PR TITLE
Adding Scala 2.13 support part 2 - pushdown

### PR DESCRIPTION
**Pull request by davidrabinowitz**
Co-authored-by: Vinay Londhe <vinaylondhe@google.com>

### CODE REVIEW

File: spark-bigquery-pushdown/spark-3.2-bigquery-pushdown_2.13/src/main/scala/com/google/cloud/spark/bigquery/pushdowns/Spark32BigQueryStrategy.scala [hunk 1]
```diff
@@ -0,0 +1,38 @@
+package com.google.cloud.spark.bigquery.pushdowns
+
+import com.google.cloud.spark.bigquery.{SparkBigQueryUtil, SupportsQueryPushdown}
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+import org.apache.spark.sql.execution.datasources.v2.DataSourceV2ScanRelation
+
+class Spark32BigQueryStrategy(expressionConverter: SparkExpressionConverter, expressionFactory: SparkExpressionFactory, sparkPlanFactory: SparkPlanFactory)
+  extends BigQueryStrategy(expressionConverter, expressionFactory, sparkPlanFactory) {
+  override def generateQueryFromPlanForDataSourceV2(plan: LogicalPlan): Option[BigQuerySQLQuery] = {
+    // DataSourceV2ScanRelation is the relation that is used in the Spark 3.1 DatasourceV2 connector
```
Comment (suryasoma) [inline spark-bigquery-pushdown/spark-3.2-bigquery-pushdown_2.13/src/main/scala/com/google/cloud/spark/bigquery/pushdowns/Spark32BigQueryStrategy.scala:10]: nit: Spark 3.2

File: spark-bigquery-pushdown/spark-3.2-bigquery-pushdown_2.13/src/main/scala/com/google/cloud/spark/bigquery/pushdowns/Spark32ExpressionConverter.scala [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+import com.google.cloud.spark.bigquery.pushdowns.SparkBigQueryPushdownUtil.blockStatement
+import org.apache.spark.sql.catalyst.expressions.{Attribute, Cast, CheckOverflow, Expression, ScalarSubquery, UnaryMinus}
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+
+/**
+ * Convert Spark 3.2 specific expressions to SQL
+ */
+class Spark32ExpressionConverter(expressionFactory: SparkExpressionFactory, sparkPlanFactory: SparkPlanFactory) extends SparkExpressionConverter() {
+  override def convertScalarSubqueryExpression(expression: Expression, fields: Seq[Attribute]): BigQuerySQLStatement = {
```
Comment (suryasoma) [inline spark-bigquery-pushdown/spark-3.2-bigquery-pushdown_2.13/src/main/scala/com/google/cloud/spark/bigquery/pushdowns/Spark32ExpressionConverter.scala:27]: shouldn't we change the corresponding method definition in https://github.com/GoogleCloudDataproc/spark-bigquery-connector/blob/master/spark-bigquery-pushdown/pushdown_common_src/main/scala/com/google/cloud/spark/bigquery/pushdowns/SparkExpressionConverter.scala#L467

File: spark-bigquery-pushdown/spark-3.2-bigquery-pushdown_2.13/src/main/scala/com/google/cloud/spark/bigquery/pushdowns/Spark32ExpressionConverter.scala [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+  override def convertUnaryMinusExpression(expression: Expression, fields: Seq[Attribute]): BigQuerySQLStatement = {
+    expression match {
+      case UnaryMinus(child, _) =>
+        ConstantString("-") +
+          blockStatement(convertStatement(child, fields))
+    }
+  }
+
+  override def convertCastExpression(expression: Expression, fields: Seq[Attribute]): BigQuerySQLStatement = {
```
Comment (suryasoma) [inline spark-bigquery-pushdown/spark-3.2-bigquery-pushdown_2.13/src/main/scala/com/google/cloud/spark/bigquery/pushdowns/Spark32ExpressionConverter.scala:55]: Am I missing something here? I couldn't find the definition for this in https://github.com/davidrabinowitz/spark-bigquery-connector/blob/pushdown213/spark-bigquery-pushdown/pushdown_common_src/main/scala/com/google/cloud/spark/bigquery/pushdowns/SparkExpressionConverter.scala

File: spark-bigquery-pushdown/spark-3.2-bigquery-pushdown_2.13/src/main/scala/com/google/cloud/spark/bigquery/pushdowns/Spark32BigQueryPushdownPlan.scala [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+import org.apache.spark.sql.execution.SparkPlan
+
+/** BigQueryPlan, with RDD defined by custom query. */
+case class Spark32BigQueryPushdownPlan(output: Seq[Attribute], rdd: RDD[InternalRow])
+  extends SparkPlan {
+
+  override def children: Seq[SparkPlan] = Nil
+
+  protected override def doExecute(): RDD[InternalRow] = {
+    doExecuteSparkPlan(output, rdd)
```
Comment (suryasoma) [inline spark-bigquery-pushdown/spark-3.2-bigquery-pushdown_2.13/src/main/scala/com/google/cloud/spark/bigquery/pushdowns/Spark32BigQueryPushdownPlan.scala:17]: Also couldn't find the implementation for this https://github.com/GoogleCloudDataproc/spark-bigquery-connector/blob/master/spark-bigquery-pushdown/pushdown_common_src/main/scala/com/google/cloud/spark/bigquery/pushdowns/SparkBigQueryPushdownUtil.scala

File: spark-bigquery-pushdown/spark-3.2-bigquery-pushdown_2.13/src/main/scala/com/google/cloud/spark/bigquery/pushdowns/Spark32ExpressionConverter.scala (hunk 2 above)
Comment (davidrabinowitz) [reply spark-bigquery-pushdown/spark-3.2-bigquery-pushdown_2.13/src/main/scala/com/google/cloud/spark/bigquery/pushdowns/Spark32ExpressionConverter.scala:27]: in PR #754

File: spark-bigquery-pushdown/spark-3.2-bigquery-pushdown_2.13/src/main/scala/com/google/cloud/spark/bigquery/pushdowns/Spark32ExpressionConverter.scala (hunk 3 above)
Comment (davidrabinowitz) [reply spark-bigquery-pushdown/spark-3.2-bigquery-pushdown_2.13/src/main/scala/com/google/cloud/spark/bigquery/pushdowns/Spark32ExpressionConverter.scala:55]: In PR #754

File: spark-bigquery-pushdown/spark-3.3-bigquery-pushdown_2.13/src/main/scala/com/google/cloud/spark/bigquery/pushdowns/Spark33BigQueryStrategy.scala [hunk 5]
```diff
@@ -0,0 +1,38 @@
+package com.google.cloud.spark.bigquery.pushdowns
+
+import com.google.cloud.spark.bigquery.{SparkBigQueryUtil, SupportsQueryPushdown}
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+import org.apache.spark.sql.execution.datasources.v2.DataSourceV2ScanRelation
+
+class Spark33BigQueryStrategy(expressionConverter: SparkExpressionConverter, expressionFactory: SparkExpressionFactory, sparkPlanFactory: SparkPlanFactory)
+  extends BigQueryStrategy(expressionConverter, expressionFactory, sparkPlanFactory) {
+  override def generateQueryFromPlanForDataSourceV2(plan: LogicalPlan): Option[BigQuerySQLQuery] = {
+    // DataSourceV2ScanRelation is the relation that is used in the Spark 3.1 DatasourceV2 connector
```
Comment (suryasoma) [inline spark-bigquery-pushdown/spark-3.3-bigquery-pushdown_2.13/src/main/scala/com/google/cloud/spark/bigquery/pushdowns/Spark33BigQueryStrategy.scala:10]: nit: Spark 3.3

**Review by suryasoma — APPROVED** [review]
LGTM

**Pull request merged**